### PR TITLE
release/4.0.0 - Including OAuth support

### PR DESCRIPTION
**Pull request by skharab-anaplan**

### CODE REVIEW

File: java/anaplan-connect-library/ac-auth/ac-auth-default/src/main/java/com/anaplan/client/auth/DeviceAuthenticator.java [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+  private static final String GRANT_TYPE = "urn:ietf:params:oauth:grant-type:device_code";
+  private static final String DEVICE_SCOPE = "openid profile email offline_access";
+  private static final String REFRESH_TOKEN = "refresh_token";
+  private static final String JKS = ".jks";
+  private final String refreshTokenKeyStorePath;
+  private final String refreshTokenKeyStoreName;
+  private final char[] keystorePass;
+  private OauthTokenInfo oauthTokenInfo;
+
+  public DeviceAuthenticator(ConnectionProperties connectionProperties,
```
Comment (skharab-anaplan) [inline java/anaplan-connect-library/ac-auth/ac-auth-default/src/main/java/com/anaplan/client/auth/DeviceAuthenticator.java:51]: Can you add Java Dc for this in ACL and here as well?

File: java/anaplan-connect-library/ac-auth/ac-auth-default/src/main/java/com/anaplan/client/auth/DeviceAuthenticator.java [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+    String keystoreDir = System.getProperty("user.home");
+    if (StringUtils.isNotBlank(System.getenv("AC_OAUTH_KEYSTORE_DIR"))) {
+      keystoreDir = System.getenv("AC_OAUTH_KEYSTORE_DIR");
+    }
+    refreshTokenKeyStorePath = keystoreDir + FileSystems.getDefault().getSeparator() + refreshTokenKeyStoreName + JKS;
+    keystorePass = CryptoUtil.encrypt(this.connectionProperties.getClientId()).toCharArray();
+  }
+
+  @Override
+  public byte[] authenticate() {
```
Comment (skharab-anaplan) [inline java/anaplan-connect-library/ac-auth/ac-auth-default/src/main/java/com/anaplan/client/auth/DeviceAuthenticator.java:68]: Can you add Java Dc for this in ACL and here as well?

File: java/anaplan-connect-library/ac-auth/ac-auth-default/src/main/java/com/anaplan/client/auth/DeviceAuthenticator.java [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+      return authToken;
+    } else {
+      synchronized (this) {
+        return refreshToken();
+      }
+    }
+  }
+
+  public void clearRefreshTokenEntry() {
+    LOG.info("Deleting already existing JKS and re-registering the device...");
```
Comment (skharab-anaplan) [inline java/anaplan-connect-library/ac-auth/ac-auth-default/src/main/java/com/anaplan/client/auth/DeviceAuthenticator.java:89]: Can you add Java Dc for this in ACL and here as well?

File: java/anaplan-connect-library/ac-auth/ac-auth-default/src/main/java/com/anaplan/client/auth/DeviceAuthenticator.java [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+      return new String(decodedKey);
+    } catch (FileNotFoundException e) {
+      LOG.error("Refresh token is not set for your client id.");
+    } catch (Exception e) {
+      LOG.error("Could not decode the refresh token or the token does not exist. {}", e.getMessage());
+    }
+    return StringUtils.EMPTY;
+  }
+
+  public OauthTokenInfo getOauthTokenInfo() {
```
Comment (skharab-anaplan) [inline java/anaplan-connect-library/ac-auth/ac-auth-default/src/main/java/com/anaplan/client/auth/DeviceAuthenticator.java:220]: Can you add Java Dc for this in ACL and here as well?

File: java/anaplan-connect-library/ac-auth/ac-auth-default/src/test/resources/test.properties [hunk 5]
```diff
@@ -0,0 +1,12 @@
+service.location=https://mock-api.anaplan.com
```
Comment (skharab-anaplan) [inline java/anaplan-connect-library/ac-auth/ac-auth-default/src/test/resources/test.properties]: Delete this file please

File: java/anaplan-connect-library/ac-impl/ac-impl-default/src/main/java/com/anaplan/client/DefaultServiceProvider.java [hunk 6]
```diff
@@ -28,5 +29,17 @@ public static Service getService(ConnectionProperties properties)
     return new Service(properties, authenticator, apiProvider);
   }
 
+  /**
+   * Get the authenticator based on properties
+   * @param properties the connection properties
+   * @return {@link DeviceAuthenticator}
+   */
+  public static DeviceAuthenticator getDeviceAuthenticator(ConnectionProperties properties) {
```
Comment (skharab-anaplan) [inline java/anaplan-connect-library/ac-impl/ac-impl-default/src/main/java/com/anaplan/client/DefaultServiceProvider.java]: Don't see this being used anywhere please check.

File: java/anaplan-connect-library/api-core/src/test/java/com/anaplan/client/listwriter/ListItemTest.java [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+    p2.setFormatMetadata(map2);
+
+    properties.add(p1);
+    properties.add(p2);
+    List<String> result = Utils.getBooleanParams(properties);
+
+    assertEquals("b", result.get(0));
+    assertEquals(1, result.size());
+  }
+}
```
Comment (skharab-anaplan) [inline java/anaplan-connect-library/api-core/src/test/java/com/anaplan/client/listwriter/ListItemTest.java]: new line end of file.

File: java/anaplan-connect-library/api-core/src/main/java/com/anaplan/client/Utils.java [hunk 8]
```diff
@@ -305,4 +319,80 @@ public static void checkTarget(final File target, String path, String file, bool
...
+    if (target == null) {
+      throw new FileNotFoundException("Path does not exist");
+    }
+    if (!target.isAbsolute()) {
+      target = Paths.get(target.toFile().getAbsolutePath());
+    }
+    return target;
+  }
+
+  public static byte[] createHash(String clientId) {
```
Comment (skharab-anaplan) [inline java/anaplan-connect-library/api-core/src/main/java/com/anaplan/client/Utils.java:356]: Add Java docs for Public methods please for all classes.

File: java/anaplan-connect-library/ac-auth/ac-auth-default/src/main/java/com/anaplan/client/auth/DeviceAuthenticator.java (hunk 1 above)
Comment (srobb-anaplan) [reply java/anaplan-connect-library/ac-auth/ac-auth-default/src/main/java/com/anaplan/client/auth/DeviceAuthenticator.java:51]: Java DC has been added

File: java/anaplan-connect-library/ac-auth/ac-auth-default/src/main/java/com/anaplan/client/auth/DeviceAuthenticator.java (hunk 2 above)
Comment (srobb-anaplan) [reply java/anaplan-connect-library/ac-auth/ac-auth-default/src/main/java/com/anaplan/client/auth/DeviceAuthenticator.java:68]: Java DC has been added

File: java/anaplan-connect-library/ac-auth/ac-auth-default/src/main/java/com/anaplan/client/auth/DeviceAuthenticator.java (hunk 3 above)
Comment (srobb-anaplan) [reply java/anaplan-connect-library/ac-auth/ac-auth-default/src/main/java/com/anaplan/client/auth/DeviceAuthenticator.java:89]: Java DC has been added

File: java/anaplan-connect-library/ac-auth/ac-auth-default/src/main/java/com/anaplan/client/auth/DeviceAuthenticator.java (hunk 4 above)
Comment (srobb-anaplan) [reply java/anaplan-connect-library/ac-auth/ac-auth-default/src/main/java/com/anaplan/client/auth/DeviceAuthenticator.java:220]: Java DC has been added

File: java/anaplan-connect-library/ac-auth/ac-auth-default/src/test/resources/test.properties (hunk 5 above)
Comment (srobb-anaplan) [reply java/anaplan-connect-library/ac-auth/ac-auth-default/src/test/resources/test.properties]: This test properties has usages across tests. In this case should I delete it?

File: java/anaplan-connect-library/ac-impl/ac-impl-default/src/main/java/com/anaplan/client/DefaultServiceProvider.java (hunk 6 above)
Comment (srobb-anaplan) [reply java/anaplan-connect-library/ac-impl/ac-impl-default/src/main/java/com/anaplan/client/DefaultServiceProvider.java]: It was not used anymore, method has been removed

File: java/anaplan-connect-library/api-core/src/test/java/com/anaplan/client/listwriter/ListItemTest.java (hunk 7 above)
Comment (srobb-anaplan) [reply java/anaplan-connect-library/api-core/src/test/java/com/anaplan/client/listwriter/ListItemTest.java]: End of line has been added

File: java/anaplan-connect-library/api-core/src/main/java/com/anaplan/client/Utils.java (hunk 8 above)
Comment (srobb-anaplan) [reply java/anaplan-connect-library/api-core/src/main/java/com/anaplan/client/Utils.java:356]: Java DC have been added across all public methods, please check latest push

**Pull request merged**